### PR TITLE
feat: improve handling ctrl+c and timeouts

### PR DESCRIPTION
**Pull request by caarlos0**
- make `ctrl+c` interrupt and exit properly
- handle timeouts with a `context.Context` and exit properly

### CODE REVIEW

File: choose/choose.go [hunk 1]
```diff
@@ -134,27 +128,13 @@ type item struct {
 }
 
 func (m model) Init() tea.Cmd {
-	return timeout.Init(m.timeout, nil)
+	return nil
 }
```
Comment (ccoVeille) [inline choose/choose.go:134]: Do you still need Init then?

File: confirm/confirm.go [hunk 2]
```diff
@@ -110,7 +103,7 @@ type model struct {
 }
 
 func (m model) Init() tea.Cmd {
-	return timeout.Init(m.timeout, m.defaultSelection)
+	return nil
 }
```
Comment (ccoVeille) [inline confirm/confirm.go:107]: Do you still need it?

File: pager/command.go [hunk 3]
```diff
@@ -40,17 +49,20 @@ func (o Options) Run() error {
...
+		tea.WithContext(ctx),
+	).Run()
 	if err != nil {
+		if errors.Is(ctx.Err(), context.DeadlineExceeded) {
+			return exit.ErrTimeout
+		}
+		if errors.Is(err, tea.ErrInterrupted) {
+			return exit.ErrAborted
+		}
 		return fmt.Errorf("unable to start program: %w", err)
```
Comment (ccoVeille) [inline pager/command.go:65]: Shouldn't this logic moved in the root command or main?

It's not logic you are copy pasting it everywhere the same dozen lines

But then it means the context should be created at top level and propagated, and not one created in each command

File: pager/pager.go [hunk 4]
```diff
@@ -28,25 +25,14 @@ type model struct {
...
 	maxWidth            int
-	timeout             time.Duration
-	hasTimeout          bool
-	timedOut            bool
 }
 
 func (m model) Init() tea.Cmd {
-	return timeout.Init(m.timeout, nil)
+	return nil
 }
```
Comment (ccoVeille) [inline pager/pager.go:32]: Remove?

**Review by ccoVeille — COMMENTED** [review]
I know it's a draft, but I looked at it

File: choose/choose.go (hunk 1 above)
Comment (caarlos0) [reply choose/choose.go]: yes

File: confirm/confirm.go (hunk 2 above)
Comment (caarlos0) [reply confirm/confirm.go]: yes - need to be a tea.Model

File: pager/command.go (hunk 3 above)
Comment (caarlos0) [reply pager/command.go]: yup, just did that :)

File: pager/pager.go (hunk 4 above)
Comment (caarlos0) [reply pager/pager.go]: can't, need to implement tea.Model

**Review by ccoVeille — APPROVED** [review]

File: choose/choose.go [hunk 5]
```diff
@@ -133,28 +127,12 @@ type item struct {
...
-func (m model) Init() tea.Cmd {
-	return timeout.Init(m.timeout, nil)
-}
+func (m model) Init() tea.Cmd { return nil }
 
 func (m model) Update(msg tea.Msg) (tea.Model, tea.Cmd) {
 	switch msg := msg.(type) {
 	case tea.WindowSizeMsg:
 		return m, nil
-	case timeout.TickTimeoutMsg:
```
Comment (aymanbagabas) [inline choose/choose.go:144]: Is this a breaking change?

File: pager/pager.go [hunk 6]
```diff
@@ -144,17 +130,14 @@ func (m model) KeyHandler(key tea.KeyMsg) (model, func() tea.Msg) {
...
 
 func (m model) View() string {
-	var timeoutStr string
-	if m.hasTimeout {
-		timeoutStr = timeout.Str(m.timeout) + " "
-	}
-	helpMsg := "\n" + timeoutStr + " ↑/↓: Navigate • q: Quit • /: Search "
+	// TODO: use help bubble here
```
Comment (aymanbagabas) [inline pager/pager.go:133]:
```suggestion
```

**Review by aymanbagabas — APPROVED** [review]
Looks good, few questions

File: pager/pager.go (hunk 6 above)
Comment (caarlos0) [reply pager/pager.go:133]: its in another PR

File: choose/choose.go (hunk 5 above)
Comment (caarlos0) [reply choose/choose.go]: shouldn't be no

**Comment by bashbunni** [comment]
This works great for the Gum commands, though when working with a script (e.g. `demo.sh`) hitting `ctrl+c` skips the current gum command, but it would be great to be able to quit the whole script with `ctrl+d`. The changes here work as expected :) would be great to hear your thoughts on how we could exit the script though without having to skip all of the remaining gum commands

**Comment by caarlos0** [comment]
@bashbunni it does work, but the script might need the `set -euo pipefail` options, e.g.:

```bash
#!/bin/bash
set -euo pipefail
time=$(gum input --placeholder "Timeout (in minutes)")
gum spin --timeout "${time}m" --show-output -- ping -c 50 google.com
echo "Done"
```

- `-e`: exits on non-zero status
- `-u`: errors on unset variables (not necessary for this example though)
- `-o pipefail`: if a command in a pipe fails, returns its status code

these options are generally good to have on any scripts as they help preventing errors.

File: choose/choose.go (hunk 5 above)
Comment (caarlos0) [reply choose/choose.go:144]: FWIW `--timeout` isn't even working in latest gum version  :P

**Pull request merged**